File: Quiz9/temp10.py
maxprofit = 0
from collections import deque

# class Queue:
#     def _init_(self):
#         self.items = deque([], 45)

#     def isEmpty(self):
#         return self.items == deque([], 45)

#     def enqueue(self, item):
#         self.items.append(item)

#     def dequeue(self):
#         return self.items.popleft()

#     def size(self):
#         return len(self.items)

class Node:
    def __init__(self, level, profit, weight):
        self.level = level
        self.profit = profit
        self.weight = weight
        self.items = []
        


def knapsack_bfs(n, p, w, W, maxprofit):
  Q = deque([], n)
  v = Node(0, 0, 0)
#   v.create_node(0,0,0)
  u = Node(0, 0, 0)
  #maxprofit = 0
  Q.append(v)
  while(len(Q) > 0):
    v = Q[0]
    Q.popleft()
    u.level = v.level
    print("u.level" + str(u.level))
    print("u.wt" + str(u.weight))
    print("u.pt" + str(u.profit))
    
    u.level = v.level
    u.weight = v.weight + w[u.level]
    u.profit = v.profit + p[u.level]
    if(u.weight <= W and u.profit > maxprofit):
      print("here")
      maxprofit = u.profit
    if(bound(u) > maxprofit):
      Q.append(u)
      print(Q)
    u.weight = v.weight
    u.profit = v.profit
    if(bound(u) > maxprofit):
      Q.append(u)
  
def bound(u):
  
  if u.weight > W:
    return 0
  else:
    result = u.profit
    j = u.level + 1
    totweight = u.weight
    while(j<n and totweight + w[j] <= W):
      totweight = totweight + w[j]
      result = result + p[j]
      j += 1
    k = j
    if k < n:
      result = result + (W - totweight)*pw[k]
    return result

n = 5
W = 13
p = [20, 30, 35, 12, 3]
w = [2, 5, 7, 3, 1]
pw = [10, 6, 5, 4, 3]
knapsack_bfs(n, p, w, W, 0)
print(maxprofit)

File: Quiz9/temp10.cpp
#include <stdio.h>
#include <iostream>
#include <limits>
#include <vector>
#include <time.h>
#include <set>
#include <cstdlib>
#include <random>
#include <queue>

using namespace std;

struct node{
    int level;
    int profit;
    int weight;
    vector<int> items;
};

int n = 5, W = 13, maxprofit = 0, cnt = 0;

int p[] = {20, 30, 35, 12, 3};
int w[] = {2, 5, 7, 3, 1};
int pw[] = {10, 6, 5, 4, 3};

float bound(node u){

    int j, k;
    int totweight = 0;
    float result = 0.0;

    if(u.weight < W){
        result = u.profit;
        j = u.level + 1;
        totweight = u.weight;

        while(totweight + w[j] <= W && j <= n){
            totweight = totweight + w[j];
            result = result + p[j];
            j++;
        }

        k = j;
        if(k <= n){
            result = result + (W - totweight) * pw[k];
        }

        return result;
    }
    
    return 0; 
}

void knapsack2(){
    queue<node> Q;
    node v;
    v.level = v.profit = v.weight = 0;
    node u;

    Q.push(v);
    v.items.push_back

    while(Q.empty() == false){
        cnt++;
        v = Q.front();
        Q.pop();
        u.level = v.level + 1;
        u.items = v.items;
        u.push_back(item[v.level-1]);
        cout << u;
        u.weight = v.weight + w[u.level - 1];
        u.profit = v.profit + p[u.level - 1];

        if(u.profit > maxprofit && u.weight <= W){
            maxprofit = u.profit;
        }

        if(bound(u) > maxprofit)
            Q.push(u);
        
        u.weight = v.weight;
        u.profit = v.profit;

        if(bound(u) > maxprofit){
            Q.push(u);
        }
    }
}

int main()
{
    knapsack2();
    cout << "Max Profit: " << maxprofit << endl;
    cout << "No. of nodes examined: " << cnt << endl;
    return 0;
}